Std/Tactic/LibrarySearch.lean

```diff
@@ -0,0 +1,530 @@
+    let newGoals ← goal.apply lem { allowSynthFailures := true }
+    try
+      act newGoals
+    catch _ =>
+      if allowFailure then
+        pure newGoals
+      else
+        failure
+
+/--
+Interleave x y interleaves the elements of x and y until one is empty and then returns
+final elements in other list.
+-/
+def interleaveWith {α β γ} (f : α → γ) (x : Array α) (g : β → γ) (y : Array β) : Array γ :=
+    Id.run do
+  let mut res := Array.mkEmpty (x.size + y.size)
+  let n := min x.size y.size
+  for h : i in [0:n] do
+    have p : i < min x.size y.size := h.2
+    have q : i < x.size := Nat.le_trans p (Nat.min_le_left ..)
+    have r : i < y.size := Nat.le_trans p (Nat.min_le_right ..)
+    res := res.push (f x[i])
+    res := res.push (g y[i])
+  let last :=
+        if x.size > n then
+          (x.extract n x.size).map f
+        else
+          (y.extract n y.size).map g
+  pure $ res ++ last
+
+/--
+Run `searchFn` on both the goal and `symm` applied to the goal.
+-/
+def librarySearchSymm (searchFn : CandidateFinder) (goal : MVarId) : MetaM (Array Candidate) := do
+  let tgt ← goal.getType
+  let l1 ← searchFn tgt
+  let coreMCtx ← getMCtx
+  let coreGoalCtx := (goal, coreMCtx)
+  if let some symmGoal ← observing? goal.applySymm then
+    let newType ← instantiateMVars  (← symmGoal.getType)
+    let l2 ← searchFn newType
+    let symmMCtx ← getMCtx
+    let symmGoalCtx := (symmGoal, symmMCtx)
+    setMCtx coreMCtx
+    pure $ interleaveWith (coreGoalCtx, ·) l1 (symmGoalCtx, ·) l2
+  else
+    pure $ l1.map (coreGoalCtx, ·)
+
+/-- A type synonym for our subgoal ranking algorithm. -/
+def SubgoalRankType := Bool × Nat × Int
+  deriving ToString
+
+instance : Ord SubgoalRankType :=
+  have : Ord (Nat × Int) := lexOrd
+  lexOrd
+
+/-- Count how many local hypotheses appear in an expression. -/
+def countLocalHypsUsed [Monad m] [MonadLCtx m] [MonadMCtx m] (e : Expr) : m Nat := do
+  let e' ← instantiateMVars e
+  return (← getLocalHyps).foldr (init := 0) fun h n => if h.occurs e' then n + 1 else n
+
+/-- Returns a tuple:
+* are there no remaining goals?
+* how many local hypotheses were used?
+* how many goals remain, negated?
+
+Larger values (i.e. no remaining goals, more local hypotheses, fewer remaining goals)
+are better.
+-/
+def subgoalRanking (goal : MVarId) (subgoals : List MVarId) : MetaM SubgoalRankType := do
+  return (subgoals.isEmpty, ← countLocalHypsUsed (.mvar goal), - subgoals.length)
+
+/-- Sort the incomplete results from `librarySearchSymm` according to
+* the number of local hypotheses used (the more the better) and
+* the number of remaining subgoals (the fewer the better).
+-/
+def sortResults (goal : MVarId) (R : Array (List MVarId × MetavarContext)) :
+    MetaM (Array (List MVarId × MetavarContext)) := do
+  let R' ← R.mapM fun (gs, ctx) => do
+    return (← withMCtx ctx (subgoalRanking goal gs), gs, ctx)
+  let R'' := R'.qsort fun a b => compare a.1 b.1 = Ordering.gt
+  return R''.map (·.2)
```

Comment:
I've done some testing and at least with `LazyDiscrTree`, `sortResults` seems to make the results worse.

I think I'll remove it for now, but may want to add it back.